src/refinement/amr_criteria.cpp

```diff
@@ -65,4 +67,39 @@ AmrTag AMRFirstDerivative::operator()(const MeshBlockData<Real> *rc) const {
   return Refinement::FirstDerivative(pmb.get(), q, refine_criteria, derefine_criteria);
 }
 
+
+AMRSecondDerivative::AMRSecondDerivative(ParameterInput *pin, std::string &block_name) {
+  field = pin->GetOrAddString(block_name, "field", "NO FIELD WAS SET");
+  if (field == "NO FIELD WAS SET") {
+    std::cerr << "Error in " << block_name << ": no field set" << std::endl;
+    exit(1);
+  }
+  refine_criteria = pin->GetOrAddReal(block_name, "refine_tol", 0.5);
+  derefine_criteria = pin->GetOrAddReal(block_name, "derefine_tol", 0.05);
+  int global_max_level = pin->GetOrAddInteger("parthenon/mesh", "numlevel", 1);
+  max_level = pin->GetOrAddInteger(block_name, "max_level", global_max_level);
+  if (max_level > global_max_level) {
```

Comment:
Not sure if we do this in the first derivative code, but might be good to hide this in a check for MPI rank zero
```suggestion
  if ((max_level > global_max_level) && (Globals::my_rank == 0)) {
```
assumes that you have
```C++
#include "globals.hpp"
```
at the top of the file.

Reply:
This too is easy enough, but honestly I'm not clear why this code has its own idea of max_level that could be <= global_max_level.

Is the idea here that you can ask for AMR refine / derefine calculations at levels less than global_max_level?  What happens then?  That is, if one does this, how do we ever end up with a level > max_level?

Reply:
I'm actually not sure---this may simply be extraneous code which should be removed.

Reply:
^which should be removed at some point. No reason you have to do it now.

Reply:
Looking at this a bit more closely, it sounds like you're suggesting that the error message written to stderr should be only done by rank 0.  But, constraining max_level <= global_max_level should be done by all ranks, right?  I'm assuming here that there will NOT be an additional synchronization step later.

Reply:
Ah, yeah that's right. I just want one error message, but the constraint should be global.

Reply:
There are use cases for which it may be desired that some refinement crit only trigger down to a certain level, which is why there's an extra check.
For example (not yet relevant for any downstream code right now as far as I'm aware of but who knows), in Enzo the gravity refinement based on particles is typically limited to several levels above the maximum level (for very deep hierarchies) so that the potential is effectively smoothed (and particles don't start acting as "black holes" on the finest levels).